libs/libc/stdio/lib_fopencookie.c

```diff
@@ -0,0 +1,229 @@
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+#include <assert.h>
+#include <errno.h>
+
+#include "libc.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct cookie_s
+{
+  FAR void *cookie;                /* Pointer to the caller's cookie struct */
+  cookie_io_functions_t cookie_io; /* Programmer-defined hook functions */
+  int cookie_fd;                   /* File descriptor */
+};
+
+static ssize_t freadcookie(FAR struct file *filep, FAR char *buf,
+                           size_t nbytes);
+static ssize_t fwritecookie(FAR struct file *filep, FAR const char *buf,
+                            size_t nbytes);
+static off_t fseekcookie(FAR struct file *filep, off_t offset, int whence);
+static int fclosecookie(FAR struct file *filep);
+
+/****************************************************************************
+  * Private Data
+ ****************************************************************************/
+
+static const struct file_operations cookie_fops =
+{
+  NULL,         /* open */
+  fclosecookie, /* close */
+  freadcookie,  /* read */
+  fwritecookie, /* write */
+  fseekcookie,  /* seek */
+  NULL,         /* ioctl */
+  NULL,         /* mmap */
+  NULL,         /* truncate */
+  NULL,         /* poll */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static ssize_t freadcookie(FAR struct file *filep, FAR char *buf,
+                           size_t nbytes)
+{
+  FAR struct cookie_s *cookie = (FAR struct cookie_s *)filep->f_priv;
+  int ret;
+
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.read != NULL);
+
+  ret = cookie->cookie_io.read(cookie->cookie, buf, nbytes);
+  return ret;
```

Comment:
Are these the user space callbacks you were referring to?  Yes, these are wrong.  This can't work this way there can be no such calls from the kernel address space.

This is similar to the way that the usrsock works.  The usrsock functions are "called" from the OS, but not directly.   Rather, the call data is marshaled through a message interface to the user code.  This is way this should be done as well.  The interface is described in include/nuttx/net/usrsock.h

At one time I wrote some user-space device drivers (character and block) based on the BSD user space drivers, but I don't think those made it into the repository.  The design was similar to the usrsock design.

Reply:
> Are these the user space callbacks you were referring to? Yes, these are wrong. This can't work this way there can be no such calls from the kernel address space.

I think `stdio` library is purely user space (and protect build builds it twice, one for kernel and one for user space) so this callback here should be correct. The issue @xiaoxiang781216 is reffering to is that the route from user call to the callback is: `fseek` from app -> `fseeko` in libc -> `lseek` in `fs/vfs` -> `fseekcookie` in libc. I think all we need to do is to avoid the usage of  `lseek` in `fs/vfs` and this way we ensure we stay in user space.